lxd/instance/drivers/qmp/commands.go

```diff
@@ -247,6 +247,85 @@ func (m *Monitor) SetMemoryBalloonSizeBytes(sizeBytes int64) error {
 	return m.run("balloon", args, nil)
 }
 
+// AddBlockDevice adds a block device.
+func (m *Monitor) AddBlockDevice(blockDev map[string]interface{}, device map[string]string) error {
+	revert := revert.New()
+	defer revert.Fail()
+
+	if blockDev != nil {
+		err := m.run("blockdev-add", blockDev, nil)
+		if err != nil {
+			return fmt.Errorf("Failed adding block device: %w", err)
+		}
+
+		revert.Add(func() {
+			blockDevDel := map[string]interface{}{
+				"node-name": blockDev["devName"],
+			}
+
+			m.run("blockdev-del", blockDevDel, nil)
+		})
+	}
+
+	if device != nil {
+		err := m.run("device_add", device, nil)
+		if err != nil {
+			return fmt.Errorf("Failed adding block device: %w", err)
+		}
+	}
+
+	revert.Success()
+	return nil
+}
+
+// RemoveBlockDevice removes a block device.
+func (m *Monitor) RemoveBlockDevice(blockDevName string, deviceID string) error {
+	if deviceID != "" {
+		deviceID := map[string]string{
+			"id": deviceID,
+		}
+
+		err := m.run("device_del", deviceID, nil)
+		if err != nil {
+			// If the device has already been removed then all good.
+			if err != nil && !strings.Contains(err.Error(), "not found") {
+				return fmt.Errorf("Failed removing device: %w", err)
+			}
+		}
+	}
+
+	if blockDevName != "" {
+		blockDevName := map[string]string{
+			"node-name": blockDevName,
+		}
+
+		var err error
+
+		// Retry a few times in case the blockdev is in use.
+		for i := 0; i < 5; i++ {
```

Comment:
I'm not a fan of this sort of arbitrary re-try logic in lower level libraries (such as this one the `qmp` package), as we can't know the situation that the caller is calling us from, and whether it is appropriate to retry or not, and for how long/how many times/how often.

Have you identified when this situation occurs? Does it only happen during hot-unplug of a running device, but not during device removal of a VM that is shutting down?

Reply:
> Have you identified when this situation occurs? Does it only happen during hot-unplug of a running device, but not during device removal of a VM that is shutting down?

I believe it's a timing issue. It happens when a disk is hot-unplugged from a running VM. This code isn't called when a VM is shutting down.

Reply:
In that case the retry logic should be pushed to where its situationally relevant (i.e `deviceDetachBlockDevice`) and out of the `qmp` package. You may want to achieve this by returning a specific error var or error type that is recognisable as a retriable error.

Reply:
Just like the wait logic in `deviceDetachNIC`.